src/v/cluster/scheduling/partition_allocator.cc

```diff
@@ -72,13 +85,177 @@ partition_allocator::allocate_partition(partition_constraints p_constraints) {
+    for (const auto& i : request.partitions) {
+        create_count += uint64_t(i.replication_factor);
+    }
+
+    uint64_t proposed_total_partitions = existent_partitions + create_count;
+
+    // Gather information about system-wide resource sizes
+    uint32_t min_core_count = 0;
+    uint64_t min_memory_bytes = 0;
+    uint64_t min_disk_bytes = 0;
+    auto all_brokers = _members.local().all_brokers();
+    for (const auto& b : all_brokers) {
+        if (min_core_count == 0) {
+            min_core_count = b->properties().cores;
+        } else {
+            min_core_count = std::min(min_core_count, b->properties().cores);
+        }
+
+        // In redpanda <= 21.11.x, available_memory_gb and available_disk_gb
+        // are not populated.  If they're zero we skip the check later.
+        if (min_memory_bytes == 0) {
+            min_memory_bytes = b->properties().available_memory_gb * 1_GiB;
+        } else if (b->properties().available_memory_gb > 0) {
+            min_memory_bytes = std::min(
+              min_memory_bytes, b->properties().available_memory_gb * 1_GiB);
+        }
+
+        if (min_disk_bytes == 0) {
+            min_disk_bytes = b->properties().available_disk_gb * 1_GiB;
+        } else if (b->properties().available_disk_gb > 0) {
+            min_disk_bytes = std::min(
+              min_disk_bytes, b->properties().available_disk_gb * 1_GiB);
+        }
+    }
+
+    // The effective values are the node count times the smallest node's
+    // resources: this avoids wrongly assuming the system will handle partition
+    // counts that only fit when scheduled onto certain nodes.
+    uint64_t effective_cpu_count = all_brokers.size() * min_core_count;
+    uint64_t effective_cluster_memory = all_brokers.size() * min_memory_bytes;
+    uint64_t effective_cluster_disk = all_brokers.size() * min_disk_bytes;
+
+    vlog(
+      clusterlog.debug,
+      "Effective cluster resources: CPU={} memory={} disk={}",
+      effective_cpu_count,
+      human::bytes(effective_cluster_memory),
+      human::bytes(effective_cluster_disk));
+
+    // Refuse to create a partition count that would violate the per-core
+    // limit.
+    const uint64_t core_limit = effective_cpu_count
+                                * allocation_node::max_allocations_per_core;
+    if (proposed_total_partitions > core_limit) {
+        vlog(
+          clusterlog.warn,
+          "Refusing to create {} partitions, exceeds core limit {}",
+          create_count,
+          effective_cpu_count * allocation_node::max_allocations_per_core);
+        return errc::topic_invalid_partitions;
+    }
+
+    // Refuse to create partitions that would violate the configured
+    // memory per partition.
+    auto memory_per_partition_replica = _memory_per_partition();
+    if (memory_per_partition_replica.has_value()) {
+        const uint64_t memory_limit = effective_cluster_memory
+                                      / memory_per_partition_replica.value();
```

Comment:
If someone sets this config knob to zero, how do we avoid crashing?

Reply:
yeh might be worth adding a validation function into config's property like was done for fallocation step.